### ISSUE TITLE
[API Proposal]: Add extension method AsReadOnly() for ICollection<T>

### ISSUE THREAD

**Issue by cakescience** [dotnet/runtime#87330]
### Background and motivation

Like `List<T>.AsReadOnly`, I'd like to propose the same for `ICollection<T>`.
If I'm not mistaken, currently to convert an instance of `ICollection<T>` to `IReadOnlyCollection<T>`, we need either:
- Call `IEnumerable.ToList<TSource>` then `List<T>.AsReadOnly` but this will allocate a new `List<T>`.
- Create a custom extension method `AsReadOnly` on `ICollection<T>` like:
```cs
public static class CollectionExtensions
{
    public static IReadOnlyCollection<T> AsReadOnly<T>(this ICollection<T> collection)
    {
        return new ReadOnlyCollectionWrapper<T>(collection);
    }

    private sealed class ReadOnlyCollectionWrapper<T> : IReadOnlyCollection<T>
    {
        private readonly ICollection<T> _collection;

        public ReadOnlyCollectionWrapper(ICollection<T> collection)
        {
            _collection = collection;
        }

        public int Count => _collection.Count;

        public IEnumerator<T> GetEnumerator()
        {
            return _collection.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
    }
}
```
- Probably other solution but nothing built-in or simple without additional memory allocation

### API Proposal

```diff
namespace System.Collections.Generic
{
    public static class CollectionExtensions
    {
+       public static IReadOnlyDictionary<T> AsReadOnly(this ICollection<T> collection);
    }
}
```

### API Usage

```cs
ICollection<int> collection = new HashSet<int>() { 1, 2, 3 };
IReadOnlyCollection<int> readonlyCollection = collection.AsReadOnly();
```

### Risks

Nothing I can see.

**Comment by huoyaoyuan** [comment]
> * Call `IEnumerable.ToList<TSource>` then `List<T>.AsReadOnly` but this will allocate a new `List<T>`.

No. This is semantically different, since it copies the content of the collection, instead of creating a read-only view.

`ReadOnlyCollection<T>` is in fact an "readonly list" wrapper. However, `HashSet<T>` implements `IReadOnlyCollection<T>` directly and there's no need to create a wrapper. You can't convert arbitrary `ICollection<T>` to `IReadOnlyCollection<T>` though, but can you give real-world business example? `ICollection<T>` itself is much less popular than `IList<T>`.

**Comment by cakescience** [comment]
I'm using a library (which I can't modify) which needs an instance of `IReadOnlyCollection<T>`. I'm currently using `ICollection<T>` and there's no easy way to convert it to `IReadOnlyCollection<T>`. Is your recommendation to switch from `ICollection<T>` to `IList<T>`?

**Comment by huoyaoyuan** [comment]
> I'm currently using `ICollection<T>` and there's no easy way to convert it to `IReadOnlyCollection<T>`.

Yes, this is a historical pain since the read-only interfaces came later.
Is the collection implemented by yourself or the BCL? You can easily add `IReadOnlyCollection<T>` to your collections. All the BCL collections should have implemented it. If not, we should probably add it.
Are you passing collections in `ICollection<T>` instead of its concrete type? This is a valid scenario. But if you can make sure the collection implements `IReadOnlyCollection<T>`, you can do a cast when passing to the third-party library as a workaround.

> Is your recommendation to switch from `ICollection<T>` to `IList<T>`?

`IReadOnlyCollection<T>` gives literally little capacity (count only) so that it's not popular. Is there any reason to not use `IList<T>`?

**Comment by cakescience** [comment]
> But if you can make sure the collection implements `IReadOnlyCollection<T>`, you can do a cast when passing to the third-party library as a workaround.

That's what I'm doing at the moment as a workaround.

> Is there any reason to not use `IList<T>`?

No particular reason. I'll switch to `IList<T>` so I can call `AsReadOnly<T>(this IList<T> list)`.

**Comment by eiriktsarpalis** [comment]
Following the pattern of other read-only collections, shouldn't this expose a new collection type akin to `ReadOnlyCollection<T>` (that is actually a read-only list)?

**Comment by huoyaoyuan** [comment]
> shouldn't this expose a new collection type akin to `ReadOnlyCollection<T>` (that is actually a read-only list)?

This is the expected approach, however the name is already taken.
